### PR TITLE
metrics: add /rafthttp/stream metrics

### PR DESCRIPTION
**Pull request by yichengq**
follower:
```
{
"cmdline": ...
"memstats": ...
"rafthttp.stream.bytes_received": {"924e2e83e93f2560": 385},
"rafthttp.stream.entries_received": {"924e2e83e93f2560": 5},
"rafthttp.stream.last_index_received": {"924e2e83e93f2560": 19214}
}
```

leader:
```
{
"cmdline": ...
"memstats": ...
"rafthttp.stream.bytes_sent": {"6e3bd23ae5f1eae0": 1001, "a8266ecf031671f3": 1001},
"rafthttp.stream.entries_sent": {"6e3bd23ae5f1eae0": 13, "a8266ecf031671f3": 13},
"rafthttp.stream.last_index_sent": {"6e3bd23ae5f1eae0": 19222, "a8266ecf031671f3": 19222}
}
```

### CODE REVIEW

**Comment by yichengq** [comment]
fixes #1766

File: etcdserver/etcdhttp/client.go [hunk 1]
```diff
@@ -284,6 +287,20 @@ func (h *statsHandler) serveLeader(w http.ResponseWriter, r *http.Request) {
 	w.Write(stats)
 }
 
+func serveStats(w http.ResponseWriter, r *http.Request) {
+	w.Header().Set("Content-Type", "application/json; charset=utf-8")
+	fmt.Fprintf(w, "{\n")
+	first := true
+	expvar.Do(func(kv expvar.KeyValue) {
```
Comment (xiang90) [inline etcdserver/etcdhttp/client.go]: add todo for getting one key or a prefix of keys

File: pkg/metrics/metrics.go [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+func (g *Gauge) Set(value int64) { g.i.Set(value) }
+
+func (g *Gauge) String() string { return g.i.String() }
+
+type nilVar struct{}
+
+func (v *nilVar) String() string { return "nil" }
+
+// Map aggregates Counters and Gauges.
+type Map struct {
```
Comment (xiang90) [inline pkg/metrics/metrics.go]: one line for consistency with the previous ones

File: etcdserver/etcdhttp/client.go (hunk 1 above)
Comment (yichengq) [inline etcdserver/etcdhttp/client.go]: added. will follow up it after this one is done

File: rafthttp/streamer.go [hunk 3]
```diff
@@ -155,12 +153,11 @@ type WriteFlusher interface {
 type streamWriter struct {
 	to   types.ID
 	term uint64
-	fs   *stats.FollowerStats
```
Comment (xiang90) [inline rafthttp/streamer.go]: we should not remove this now. we can remove it when we want to deprecate /v2/stats

Comment (yichengq) [inline rafthttp/streamer.go]: k

**Comment by yichengq** [comment]
updated. @xiang90

**Comment by yichengq** [comment]
I add `GetOrNewMap` func to make it easier to init.
I don't change the `NewCounter` public function. It is for one metrics that uses counter only, which I assume is the most common case. For example, `store.operation_total`, `store.operation_failed`.

**Comment by yichengq** [comment]
@xiang90 refreshed. Is it good now?

**Comment by xiang90** [comment]
lgtm

**Pull request merged**